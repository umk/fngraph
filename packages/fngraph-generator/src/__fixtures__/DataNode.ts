import DataNode from '../DataNode'

export function createDataNode(overrides?: Partial<DataNode>): DataNode {
  return {
    incoming: [],
    outgoing: [],
    getter: jest.fn(),
    priority: 0,
    invert: false,
    getProperties: () => [],
    ...overrides,
  }
}
