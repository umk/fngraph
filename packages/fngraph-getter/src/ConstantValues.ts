import { DataRecord } from '@fngraph/data'
import {
  CONTEXT_GROUP_INVERSION,
  Falsy,
  GeneratorValue,
  Getter,
  GetterFactory,
} from '@fngraph/generator'

import { createIteration } from './IterationFunction'

type ConstantValues<V extends DataRecord> = Array<V | Falsy> | V | Falsy

export function createConstant<V extends DataRecord>(
  source: ConstantValues<V>,
): GetterFactory<never, V> {
  return function (_incoming, outgoing): Getter {
    const iterationF = createIteration(outgoing)
    return async function* (
      _properties,
      records,
      context,
      getRecord,
      invert,
    ): AsyncGenerator<GeneratorValue> {
      const iteration = iterationF(context, invert)
      for await (const current of records) {
        context.advanceToValueGroup(current)
        const base = getRecord(current.contexts)
        if (base) {
          const valuesArr = Array.isArray(source) ? source : [source]
          if (!invert) {
            yield* iteration(current, base, valuesArr.filter(Boolean) as Array<V>)
          } else {
            if (!valuesArr.some(Boolean)) {
              yield* iteration(current, base, CONTEXT_GROUP_INVERSION as V)
            }
          }
        }
      }
    }
  }
}

export default ConstantValues
