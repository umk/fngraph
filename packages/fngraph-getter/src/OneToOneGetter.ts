import { DataRecord } from '@fngraph/data'
import { ContextGroup, Falsy, GeneratorValue, Getter, GetterFactory } from '@fngraph/generator'

import { createInversion } from './InversionFunction'
import { createIteration } from './IterationFunction'

export type OneToOneGetter<P extends DataRecord, R extends DataRecord> = (
  record: P,
) => Promise<R | Falsy>

export function createOneToOne<P extends DataRecord, R extends DataRecord>(
  source: OneToOneGetter<P, R>,
): GetterFactory<P, R> {
  return function (incoming, outgoing): Getter {
    const iterationF = createIteration(outgoing)
    return async function* (records, context, getRecord, invert): AsyncGenerator<GeneratorValue> {
      const inversion = createInversion(invert)
      const iteration = iterationF(context, invert)
      async function getSourceRecord(current: GeneratorValue) {
        if (context.groups.length > context.groupIndex) {
          const derived = context.groups[context.groupIndex] as ContextGroup<R> | undefined
          if (derived) {
            // The cached derived record is defined only if and only
            // if the context is also defined.
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            const base = getRecord(current.contexts)!
            return { base, derived }
          }
        } else {
          const base = getRecord(current.contexts)
          if (base) {
            const derived = inversion(await source(incoming(base)))
            if (context.isCached) context.groups.push(derived || undefined)
            if (derived) return { base, derived }
          } else {
            context.groups.push(undefined)
          }
        }
        return undefined
      }
      for await (const current of records) {
        context.advanceToValueGroup(current)
        const record = await getSourceRecord(current)
        if (record) {
          yield* iteration(current, record.base, record.derived as R)
        }
        context.groupIndex++
      }
    }
  }
}

export default OneToOneGetter
