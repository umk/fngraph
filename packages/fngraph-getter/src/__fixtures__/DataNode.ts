import { DataNode } from '@fngraph/generator'

export function createDataNode(overrides?: Partial<DataNode>): DataNode {
  return {
    incoming: [],
    outgoing: [],
    getter: jest.fn(),
    priority: 0,
    invert: false,
    getProperties: () => [],
    ...overrides,
  }
}
