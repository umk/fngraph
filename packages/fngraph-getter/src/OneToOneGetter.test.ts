import { DataRecord, DeclarationID } from '@fngraph/data'
import {
  contextAsRecord,
  createDataRecordGenerator,
  DataNodeSequence,
  recordAsContext,
} from '@fngraph/generator'

import { createConstant } from './ConstantValues'
import { createOneToOne } from './OneToOneGetter'
import { createDataNode } from './__fixtures__'

describe('createOneToOne', () => {
  const declarationA = 'declaration A' as DeclarationID
  const declarationB = 'declaration B' as DeclarationID
  const declarationC = 'declaration C' as DeclarationID
  const declarationD = 'declaration D' as DeclarationID
  it('iterates through getter records', async () => {
    const nodeA = createDataNode({
      outgoing: [declarationA, declarationB],
      getter: createConstant([
        { [declarationA]: 1, [declarationB]: 2 },
        { [declarationA]: 3, [declarationB]: 4 },
      ])(contextAsRecord, recordAsContext),
    })
    const nodeB = createDataNode({
      incoming: [declarationA, declarationB],
      outgoing: [declarationC],
      getter: createOneToOne((record) => {
        const valueA = record[declarationA] as number
        const valueB = record[declarationB] as number
        return Promise.resolve({ [declarationC]: valueA + valueB })
      })(contextAsRecord, recordAsContext),
    })
    const sequence = DataNodeSequence.create([nodeA, nodeB])
    const generator = createDataRecordGenerator(sequence, [])
    const records: Array<DataRecord> = []
    for await (const record of generator()) records.push(record)
    expect(records).toMatchSnapshot()
  })
  it('permutates records with another node', async () => {
    /**
     * Must result in the following evaluation sequence:
     *
     *   (B) <- (C) <- (A)
     *
     *        - or -
     *
     *   (B) <- (A) <- (C)
     */
    const nodeA = createDataNode({
      outgoing: [declarationA, declarationB],
      getter: createConstant([
        { [declarationA]: 1, [declarationB]: 2 },
        { [declarationA]: 3, [declarationB]: 4 },
      ])(contextAsRecord, recordAsContext),
    })
    const getter = jest.fn((record) => {
      const valueA = record[declarationA] as number
      const valueB = record[declarationB] as number
      return Promise.resolve({ [declarationC]: valueA + valueB })
    })
    const nodeB = createDataNode({
      incoming: [declarationA, declarationB],
      outgoing: [declarationC],
      getter: createOneToOne(getter)(contextAsRecord, recordAsContext),
    })
    const nodeC = createDataNode({
      outgoing: [declarationD],
      getter: createConstant([{ [declarationD]: 10 }, { [declarationD]: 20 }])(
        contextAsRecord,
        recordAsContext,
      ),
    })
    const sequence = DataNodeSequence.create([nodeA, nodeB, nodeC])
    const generator = createDataRecordGenerator(sequence, [])
    const records: Array<DataRecord> = []
    for await (const record of generator()) records.push(record)
    expect(records).toMatchSnapshot()
    expect(getter).toHaveBeenCalledTimes(2)
  })
})
