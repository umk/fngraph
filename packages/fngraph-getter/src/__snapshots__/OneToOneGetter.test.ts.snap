// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createOneToOne iterates through getter records 1`] = `
[
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 3,
  },
  {
    "declaration A": 3,
    "declaration B": 4,
    "declaration C": 7,
  },
]
`;

exports[`createOneToOne permutates records with another node 1`] = `
[
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 3,
    "declaration D": 10,
  },
  {
    "declaration A": 3,
    "declaration B": 4,
    "declaration C": 7,
    "declaration D": 10,
  },
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 3,
    "declaration D": 20,
  },
  {
    "declaration A": 3,
    "declaration B": 4,
    "declaration C": 7,
    "declaration D": 20,
  },
]
`;
